<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <!-- Required launch arguements -->
  <arg name="num_robots" description="Number of robots spawned and used in the experiment"/>

  <!-- Optional launch arguments -->
  <arg name="seed"       default="0" description="Random seed used for randomly initializing vehicle poses"/>
  <arg name="world_name" default="sydney_regatta"
       description="String name of Gazebo world, as present on the &lt;world&gt; attribute of the loaded SDF data" />

  <!-- Generated launch arguments -->
  <arg name="init_poses"
       default="$(command '$(exec-in-pkg experiment_setup virelex)
		--num-robots $(var num_robots)
		--seed $(var seed)
		init_poses' fail)"
       description="2D initial pose of spawned experiment vehicles, encoded as a semicolon-delimitted list x,y,theta triples"/>
  <arg name="goal_poses"
       default="$(command '$(exec-in-pkg experiment_setup virelex)
		--num-robots $(var num_robots)
		--seed $(var seed)
		goal_poses' fail)"
       description="2D goal pose of spawned experiment vehicles, encoded as a semicolon-delimitted list of x,y pairs"/>


  <!-- Experiment action planner arguments -->
  <arg name="method" default="RL" description="Method to used for action planner">
    <choice value="RL"/>
    <choice value="MPC"/>
    <choice value="APF"/>
  </arg>
  <arg name="agent_type" default="AC-IQN" description="RL agent type used (only applied when method=='RL'">
    <!-- Continuous-control agents -->
    <choice value="AC-IQN"/>
    <choice value="DDPG"/>
    <choice value="SAC"/>
    <!-- Discrete-control agents -->
    <choice value="IQN"/>
    <choice value="DQN"/>
    <choice value="Rainbow"/>
  </arg>
  <!-- TODO: properly associate the default trained models with `agent_type` parameter above -->
  <arg name="model_path" default="$(find-pkg-share virelex)/trained/traced_AC_IQN_model.pt"/>

  <!-- Experiment manager node;  handles unpause/pause on collision/timeout/completion condition -->
  <node pkg="virelex" exec="experiment_manager">
    <param name="num_robots" value="$(var num_robots)" />
    <param name="world_name" value="$(var world_name)" />
  </node>

  <!-- Main VRX competition launch file -->
  <include file="$(find-pkg-share virelex)/launch/competition.launch.py">
    <arg name="init_poses" value="$(var init_poses)" />
    <arg name="goals"      value="$(var goal_poses)" />

    <arg name="method"     value="$(var method)"    />
    <arg name="agent_type" value="$(var agent_type)"/>
    <arg name="model_path" value="$(var model_path)"/>

    <arg name="world" value="$(var world_name)"/>
  </include>
</launch>
